Fix overflow possibility triggered by pathological (e.g., clipping) audio

dbry opened the pull request:
@hayguen ,this is what I came up with for a fix to the overflow. The extra time to calculate the sum is negligible compared to to the other times, and in the stereo and "fast" case we're looping through all the samples anyway so it's almost free. This also allows the best possible precision for the actual audio (usually much better than before). And I included the check for silence (which also provides a great speedup for audio that contains runs of silence).

Assuming I didn't miss anything I'd like to go with this first as it seems like your other changes are (at least mostly) cosmetic. I have some myself and some other improvements, so I'll work on those afterward.

Thanks!

hayguen commented:
@dbry , having a quick look over your commit
FYI: i modified my patch meanwhile, reducing it to 32 bit, and force pushed it.

special case of zeros:
* you are now catching the case, that all (`cnxt->longest`) samples are zero.
* but this doesn't cover the case, when sequence starts with only `longest / 2` zeros. in these cases, the period length of amount of zeros will always win for best_period. not sure, if this is good and desired? it might be ok

overflows/clipping:
* `scaler` fixes overflows at calculation of `sum * scaler`
* there are still overflows when calculating `high_side_diff` and also `low_side_diff`: the difference can increase it's result
* now, i also realize that the type for `side_diff`s need to be `unsigned`, cause the differences' range is 0 .. MAX. think, i had mis-interpreted type/sign converted values as overflow
* but for sure, `high_side_diff * 2` in `if (low_side_diff > high_side_diff * 2)` and also `low_side_diff * 2` will overflow
* all the follow-up overflow places had lead me to use `UINT32_MAX / 4` as maximum value in `results[]`. meanwhile thinking, that `UINT32_MAX / 2` might be sufficient ..

i'm at home now - not having the pathological file here, which produced the UBsan error. i can't test before next Tuesday.

dbry commented:
> @dbry , having a quick look over your commit FYI: i modified my patch meanwhile, reducing it to 32 bit, and force pushed it.
> 
> special case of zeros:
> 
> * you are now catching the case, that all (`cnxt->longest`) samples are zero.
> * but this doesn't cover the case, when sequence starts with only `longest / 2` zeros. in these cases, the period length of amount of zeros will always win for best_period. not sure, if this is good and desired? it might be ok

Yes, I think this is the best behavior. We get rid of the silent samples and then start fresh with non-zero audio.

> 
> overflows/clipping:
> 
> * `scaler` fixes overflows at calculation of `sum * scaler`
> * there are still overflows when calculating `high_side_diff` and also `low_side_diff`: the difference can increase it's result
> * now, i also realize that the type for `side_diff`s need to be `unsigned`, cause the differences' range is 0 .. MAX. think, i had mis-interpreted type/sign converted values as overflow

This is definitely an edge-case because ordinarily the diffs between adjacent periods will be very small, but changing this to unsigned int is the correct thing to do.

> * but for sure, `high_side_diff * 2` in `if (low_side_diff > high_side_diff * 2)` and also `low_side_diff * 2` will overflow

Fixed this by using dividing by two on the other side instead.

> * all the follow-up overflow places had lead me to use `UINT32_MAX / 4` as maximum value in `results[]`. meanwhile thinking, that `UINT32_MAX / 2` might be sufficient ..

Now I think UINT32_MAX - 1 is sufficient. The -1 is just to make sure that the case for diff == 0 (which gives UINT32_MAX) always wins.

> 
> i'm at home now - not having the pathological file here, which produced the UBsan error. i can't test before next Tuesday.

I created a artificial file to trigger the error and am pretty convinced that I have covered all the possible cases now.

However, I was curious how you triggered an UB error because unsigned int overflow is not actually UB (although in this application it _is_ a bug). I was only able to get this to trigger by adding `-fsanitize=unsigned-integer-overflow`. I assume that's what you used too, right?

hayguen commented:
@dbry :

> Yes, I think this is the best behavior. We get rid of the silent samples and then start fresh with non-zero audio.

ok.

> This is definitely an edge-case because ordinarily the diffs between adjacent periods will be very small, but changing this to unsigned int is the correct thing to do.

full ACK

> Fixed this by using dividing by two on the other side instead.

absolutely right. had meanwhile realized the same .. and thought: OMG.
you were slightly quicker to post

> Now I think UINT32_MAX - 1 is sufficient. The -1 is just to make sure that the case for diff == 0 (which gives UINT32_MAX) always wins.

ACK

> However, I was curious how you triggered an UB error because unsigned int overflow is not actually UB (although in this application it is a bug). I was only able to get this to trigger by adding -fsanitize=unsigned-integer-overflow. I assume that's what you used too, right?

i had tried this night with following test file - after converting to 48 kHz and WAVe format:
https://commons.wikimedia.org/wiki/File:NATO_Phonetic_Alphabet_reading.ogg

i compiled with `gcc -O0 -g -fsanitize=undefined *.c -lm -o audio-stretch`

depending on the branch, i started and got:

```
./audio-stretch  -r0.75 -y NATO_Phonetic_Alphabet_reading.wav NATO_Phonetic_Alphabet_reading_ut_0-75.wav

 AUDIO-STRETCH  Time Domain Harmonic Scaling Demo  Version 0.2
 Copyright (c) 2019 David Bryant. All Rights Reserved.

stretch.c:405:44: runtime error: signed integer overflow: -2147483648 * 2 cannot be represented in type 'int'
stretch.c:407:49: runtime error: signed integer overflow: -1431655768 * 2 cannot be represented in type 'int'
```

i haven't tested your last commit .. but it looks quite same to my local change.
think, you can go with this. i'm sure it's good.

dbry commented:
@hayguen 
> > However, I was curious how you triggered an UB error because unsigned int overflow is not actually UB (although in this application it is a bug). I was only able to get this to trigger by adding -fsanitize=unsigned-integer-overflow. I assume that's what you used too, right?
> 
> i had tried this night with following test file - after converting to 48 kHz and WAVe format: https://commons.wikimedia.org/wiki/File:NATO_Phonetic_Alphabet_reading.ogg
> 
> i compiled with `gcc -O0 -g -fsanitize=undefined *.c -lm -o audio-stretch`
> 
> depending on the branch, i started and got:
> 
> ```
> ./audio-stretch  -r0.75 -y NATO_Phonetic_Alphabet_reading.wav NATO_Phonetic_Alphabet_reading_ut_0-75.wav
> 
>  AUDIO-STRETCH  Time Domain Harmonic Scaling Demo  Version 0.2
>  Copyright (c) 2019 David Bryant. All Rights Reserved.
> 
> stretch.c:405:44: runtime error: signed integer overflow: -2147483648 * 2 cannot be represented in type 'int'
> stretch.c:407:49: runtime error: signed integer overflow: -1431655768 * 2 cannot be represented in type 'int'
> ```

Ah, this makes sense now. You uncovered the problem with `xxxx_side_diff * 2` overflowing. The other problem with `(sum * 128)` overflowing could _not_ have been uncovered this way because it was unsigned integers and therefore not UB, but was a more serious problem because it could have caused missing the best period entirely (instead of just off-by-one), and also affected the normal as well as fast modes. It also was only an issue with very loud samples and not easily triggered.

Funny though you should be using that sample as I am very familiar with it! I was doing work for a company that was embedding the spoken NATO alphabet in a device, and that was the sample they provided me. They decided against it because the accent was too British (it sounds like Keira Knightley to me) and so I recorded the samples myself (I am from California and therefore have no accent :smiley: ). I also created my [ADPCM](https://github.com/dbry/adpcm-xq) encoder for that project.

> 
> i haven't tested your last commit .. but it looks quite same to my local change. think, you can go with this. i'm sure it's good.

Thanks!
